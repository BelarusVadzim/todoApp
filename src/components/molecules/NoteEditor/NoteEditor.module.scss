@import "src/variables.scss";

$active-text-color: $primary-font-color;
$completed-text-color: $secondary-font-color;
$item-border-color: $inactive-color;

.noteEditor {
  height: 3rem;
  width: 100%;
  vertical-align: middle;
  display: flex;
  justify-content: start;
}

.leftField {
  min-width: 4rem;
  display: flex;
}

.rightField {
  min-width: 4rem;
  display: flex;
}


