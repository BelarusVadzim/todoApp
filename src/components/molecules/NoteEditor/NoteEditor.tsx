import React from 'react';
import { Textbox } from 'components/atoms';
import style from './NoteEditor.module.scss';
import { PropsWithClassName, TodoNote } from 'types';
import { useTodoStateService } from 'hooks';

type NoteEditorProps = PropsWithClassName & {
  note?: TodoNote;
};

const NoteEditor: React.FC<NoteEditorProps> = ({ className = '', note }) => {
  const combinedClassNames = `${style.noteEditor} ${className}`;

  const { dispatchTodoItemCreated, dispatchTodoItemEdited } =
    useTodoStateService();

  const editFinished = (text: string) => {
    if (text)
      if (note)
        dispatchTodoItemEdited({ ...note, text });
      else dispatchTodoItemCreated({ text });
  };

  return (
    <div className={combinedClassNames}>
      <div className={style.leftField} />
      <Textbox onReturn={editFinished} />
      <div className={style.rightField} />
    </div>
  );
};

export default NoteEditor;
