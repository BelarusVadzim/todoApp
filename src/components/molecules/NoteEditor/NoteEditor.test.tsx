import { ChangeEvent } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import NoteEditor from './NoteEditor';

const dispatchTodoItemCreated = jest.fn();
const dispatchTodoItemEdited = jest.fn();

jest.mock('./NoteEditor.module.scss', () => ({
  noteEditor: 'noteEditorStyle',
  leftField: 'leftFieldStyle',
  rightField: 'rightFieldStyle',
}));

jest.mock('hooks', () => ({
  useTodoStateService: () => ({
    dispatchTodoItemCreated,
    dispatchTodoItemEdited,
  }),
}));

type MockTextboxProps = {
  onReturn?: (text: string) => void;
  test?: string;
};

jest.mock('components/atoms', () => ({
  Textbox: jest.fn(({ onReturn, test }: MockTextboxProps) => {
    const onChangeHandler = (arg: ChangeEvent<HTMLInputElement>) => {
      if (onReturn) {
        onReturn(arg.target.value);
      }
    };

    return <input type="text" onChange={onChangeHandler} id={test} />;
  }),
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('<NoteEditor />', () => {
  it('should render properly', () => {
    const component = render(<NoteEditor />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('when className specified should render properly', () => {
    const component = render(<NoteEditor className="extraClass" />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('when edit exited note finished should call expected functions with expected values', () => {
    render(<NoteEditor note={{ text: '' }} />);

    const inputElement = screen.getByRole('textbox');

    userEvent.type(inputElement, '1{Enter}');

    expect(dispatchTodoItemCreated).toBeCalledTimes(0);
    expect(dispatchTodoItemEdited).toBeCalledTimes(1);
  });

  it('when edit new note finished should call expected functions with expected values', () => {
    render(<NoteEditor />);

    const inputElement = screen.getByRole('textbox');

    userEvent.type(inputElement, '1{Enter}');

    expect(dispatchTodoItemCreated).toBeCalledTimes(1);
    expect(dispatchTodoItemEdited).toBeCalledTimes(0);
  });
});
