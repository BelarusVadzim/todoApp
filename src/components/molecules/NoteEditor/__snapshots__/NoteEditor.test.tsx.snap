// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteEditor /> should render properly 1`] = `
<DocumentFragment>
  <div
    class="noteEditorStyle "
  >
    <div
      class="leftFieldStyle"
    />
    <input
      type="text"
    />
    <div
      class="rightFieldStyle"
    />
  </div>
</DocumentFragment>
`;

exports[`<NoteEditor /> when className specified should render properly 1`] = `
<DocumentFragment>
  <div
    class="noteEditorStyle extraClass"
  >
    <div
      class="leftFieldStyle"
    />
    <input
      type="text"
    />
    <div
      class="rightFieldStyle"
    />
  </div>
</DocumentFragment>
`;
